### ISSUE TITLE
chap03_Linear_Regression_regression_model2

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#23]
# -*- coding: utf-8 -*-
"""
y변수 : 1칼럼
x변수 : 2~4칼럼
model 최적화 알고리즘 : GD -> Adam
"""

import tensorflow.compat.v1 as tf # ver1.x
tf.disable_v2_behavior() # ver2.x 사용 안함
import pandas as pd # csv file read
from sklearn.model_selection import train_test_split # split
from sklearn.metrics import mean_squared_error # model 평가

### 1. 공급 data 생성
iris = pd.read_csv("C:\\ITWILL\\6_Tensorflow\\data/iris.csv")

cols = list(iris.columns)
x_data = iris[cols[1:4]] # 2~4칼럼
y_data = iris[cols[0]]
x_data.shape # (150, 3)
y_data.shape # (150,)

x_train, x_test, y_train, y_test = train_test_split(
    x_data, y_data, test_size = 0.3)

x_train.shape # (105, 3)
x_test.shape # (45, 3)

### 2. X, Y 변수 정의 : 공급형 변수
-train과 test 가 길이가 달라서 가변길이 none으로 변수 지정
X = tf.placeholder(dtype=tf.float32, shape = [None, 3]) # [관측치, 입력수]
Y = tf.placeholder(dtype=tf.float32, shape = [None]) # [관측치]

### 3. a(w), b 변수 정의 : 난수 초기값
a = tf.Variable(tf.random_normal(shape = [3,1])) # [입력수, 출력수]
b = tf.Variable(tf.random_normal(shape = [1])) # [출력수]

### 4. model 생성
model = tf.matmul(X, a) + b # 예측치

loss = tf.reduce_mean(tf.square(Y - model))

opt = tf.train.AdamOptimizer(0.8) # 학습율 = 0.5

train = opt.minimize(loss) # 소실 최소화 식

### 5. model 학습 : model 최적화(최적의 a, b update)
with tf.Session() as sess :
    sess.run(tf.global_variables_initializer()) # a,b 초기화
    a_val, b_val = sess.run([a,b])
    print("최초 기울기와 절편 : a = {}, b ={}".format(a_val, b_val))
    
    - 훈련용 공급 data
    feed_data = {X : x_train, Y : y_train}
    
    - 반복학습 100회
    for step in range(200) : 
        _, loss_val = sess.run([train, loss], feed_dict = feed_data) # 사용하지 않으면 _로 받기
        print("step = {}, loss = {:.5f}".format(step+1, loss_val))
        
    model 최적화
    a_up, b_up = sess.run([a,b])
    print("수정된 기울기와 절편 : a = {}, b ={}".format(a_up, b_up))
    '''
    수정된 기울기와 절편 : a = [[ 0.6009913]
                               [ 0.4524133]
                               [-0.8387338]], b =[3.3090477]
    '''
    
    - 테스트용 공급 data
    feed_data_test = {X : x_test, Y : y_test}
    
    - Y(정답) vs model(예측치)
    y_true = sess.run(Y, feed_dict = feed_data_test) # y만걸러서 공급
    y_pred = sess.run(model, feed_dict = feed_data_test) # x만 걸러서 공급
    
    - 모델 평가
    mse = mean_squared_error(y_true, y_pred)
    print("mse=",mse)     # mse= 0.48795688
    
'''
1차 - 학습률 0.5 반복횟수 100회 mse= 0.48795688
2차 - 학습률 0.4 반복횟수 100회 mse= 0.59019107
3차 - 학습률 0.4 반복횟수 200회 mse= 0.52500373
4차 - 학습률 0.1 반복횟수 200회 mse= 0.60548925
5차 - 학습률 0.8 반복횟수 200회 mse= 0.6607265
'''
